src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.Transcoding.cs

```diff
@@ -0,0 +1,1452 @@
+                                    }
+
+                                    goto ProcessTwoByteSequenceSkipOverlongFormCheck;
+                                }
+                            }
+
+                            // If we reached this point, the next sequence is something other than a valid
+                            // two-byte sequence, so go back to the beginning of the loop.
+                            goto AfterReadDWord;
+                        }
+                        else
+                        {
+                            goto ProcessRemainingBytesSlow; // Running out of data - go down slow path
+                        }
+                    }
+
+                    // The buffer contains a 2-byte sequence followed by 2 bytes that aren't a 2-byte sequence.
+                    // Unlikely that a 3-byte sequence would follow a 2-byte sequence, so perhaps remaining
+                    // bytes are ASCII?
+
+                    uint charToWrite = ExtractCharFromFirstTwoByteSequence(thisDWord); // optimistically compute this now, but don't store until we know dest is large enough
+
+                    if (UInt32ThirdByteIsAscii(thisDWord))
+                    {
+                        if (UInt32FourthByteIsAscii(thisDWord))
+                        {
+                            if (outputCharsRemaining < 3)
+                            {
+                                goto ProcessRemainingBytesSlow; // running out of output buffer
+                            }
+
+                            pOutputBuffer[0] = (char)charToWrite;
+                            if (BitConverter.IsLittleEndian)
+                            {
+                                thisDWord >>= 16;
+                                pOutputBuffer[1] = (char)(byte)thisDWord;
+                                thisDWord >>= 8;
+                                pOutputBuffer[2] = (char)thisDWord;
+                            }
+                            else
+                            {
+                                pOutputBuffer[2] = (char)(byte)thisDWord;
+                                pOutputBuffer[1] = (char)(byte)(thisDWord >> 8);
+                            }
+                            pInputBuffer += 4;
+                            pOutputBuffer += 3;
+                            outputCharsRemaining -= 3;
+
+                            continue; // go back to original bounds check and check for ASCII
+                        }
+                        else
+                        {
+                            if (outputCharsRemaining < 2)
+                            {
+                                goto ProcessRemainingBytesSlow; // running out of output buffer
+                            }
+
+                            pOutputBuffer[0] = (char)charToWrite;
+                            pOutputBuffer[1] = (char)(byte)(thisDWord >> (BitConverter.IsLittleEndian ? 16 : 8));
```

Comment:
Swap order, to avoid one bound-check. I.e.
```c#
pOutputBuffer[1] = (char)(byte)(thisDWord >> (BitConverter.IsLittleEndian ? 16 : 8));
pOutputBuffer[0] = (char)charToWrite;
```

Reply:
Ups, it's a pointer, so please ignore that comment.